File: packages/php/src/Seguranca/Ataques/AtaqueCusum.php
<?php

namespace Application\Seguranca\Ataques;

use Application\Seguranca\AlvoCriptografico;
use Application\Seguranca\AtaqueInterface;
use Application\Seguranca\ResultadoAtaque;
use Application\Seguranca\SkipAtaqueException;

/**
 * Somas cumulativas (cusum), teste do NIST SP800-22. Converte os bits em
 * passos +1/-1 e observa o maior desvio da caminhada aleatória. Um viés
 * pequeno que o monobit quase não vê faz o desvio acumulado crescer.
 * Para uma sequência aleatória, max|S|/sqrt(n) fica tipicamente abaixo de 3.
 */
class AtaqueCusum implements AtaqueInterface
{
    public function __construct(private int $tamanho = 8192, private float $limiteZ = 4.0)
    {
    }

    public function nome(): string
    {
        return 'Somas cumulativas (cusum)';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        $ks = $alvo->gerarKeystreamBruto($alvo->chaveDeTeste(), random_bytes($alvo->tamanhoIv()), 'enc', $this->tamanho);
        if ($ks === null) {
            throw new SkipAtaqueException('Alvo não expõe gerarKeystreamBruto.');
        }

        $soma = 0;
        $maxAbs = 0;
        for ($i = 0; $i < strlen($ks); $i++) {
            $byte = ord($ks[$i]);
            for ($b = 7; $b >= 0; $b--) {
                $soma += (($byte >> $b) & 1) === 1 ? 1 : -1;
                $maxAbs = max($maxAbs, abs($soma));
            }
        }
        $n = strlen($ks) * 8;
        $z = $maxAbs / sqrt($n);
        $vulneravel = $z > $this->limiteZ;

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: $vulneravel,
            severidade: $vulneravel ? 'media' : 'info',
            detalhes: sprintf('max|S|=%d sobre %d bits, max|S|/sqrt(n)=%.2f (limite=%.1f)', $maxAbs, $n, $z, $this->limiteZ),
            dados: ['max_abs' => $maxAbs, 'z' => $z],
        );
    }
}

File: packages/julia/src/Seguranca/Ataques/AtaqueCusum.jl
"""
Somas cumulativas (cusum), teste do NIST SP800-22. Converte os bits em
passos +1/-1 e observa o maior desvio da caminhada aleatória. Um viés
pequeno que o monobit quase não vê faz o desvio acumulado crescer.
Para uma sequência aleatória, max|S|/sqrt(n) fica tipicamente abaixo de 3.
"""
struct AtaqueCusum <: Ataque
    tamanho::Int
    limite_z::Float64
end

AtaqueCusum(; tamanho::Integer = 8192, limite_z::Real = 4.0) =
    AtaqueCusum(Int(tamanho), Float64(limite_z))

nome(::AtaqueCusum) = "Somas cumulativas (cusum)"

function executar(a::AtaqueCusum, alvo::AlvoCriptografico)
    ks = gerar_keystream_bruto(
        alvo,
        chave_de_teste(alvo),
        random_bytes(tamanho_iv(alvo)),
        "enc",
        a.tamanho,
    )

    soma = 0
    max_abs = 0
    for byte in ks
        for b in 7:-1:0
            soma += ((byte >> b) & 1) == 1 ? 1 : -1
            max_abs = max(max_abs, abs(soma))
        end
    end
    n = length(ks) * 8
    z = max_abs / sqrt(n)
    vulneravel = z > a.limite_z

    return ResultadoAtaque(
        nome(a),
        vulneravel,
        vulneravel ? "media" : "info",
        "max|S|=$max_abs sobre $n bits, max|S|/sqrt(n)=$(fmt(z, 2)) (limite=$(fmt(a.limite_z, 1)))",
        Dict{String,Any}("max_abs" => max_abs, "z" => z),
    )
end
